presto-tests/src/test/java/com/facebook/presto/memory/TestMemoryManager.java

```diff
@@ -62,6 +65,77 @@
+                .put("task.operator-pre-allocated-memory", "0B")
+                .put("query.kill-on-out-of-memory-delay", "5s")
+                .put("query.kill-on-out-of-memory", "true")
+                .build();
+
+        try (DistributedQueryRunner queryRunner = createQueryRunner(TINY_SESSION, properties)) {
+            // Reserve all the memory
+            QueryId fakeQueryId = new QueryId("fake");
+            for (TestingPrestoServer server : queryRunner.getServers()) {
+                for (MemoryPool pool : server.getLocalMemoryManager().getPools()) {
+                    assertTrue(pool.tryReserve(fakeQueryId, pool.getMaxBytes()));
+                }
+            }
+
+            List<Future<?>> queryFutures = new ArrayList<>();
+            for (int i = 0; i < 2; i++) {
+                queryFutures.add(executor.submit(() -> queryRunner.execute("SELECT COUNT(*), clerk FROM orders GROUP BY clerk")));
+            }
+
+            ClusterMemoryManager memoryManager = queryRunner.getCoordinator().getClusterMemoryManager();
+            ClusterMemoryPool reservedPool;
+            while ((reservedPool = memoryManager.getPools().get(RESERVED_POOL)) == null) {
+                MILLISECONDS.sleep(10);
+            }
+
+            ClusterMemoryPool generalPool = memoryManager.getPools().get(GENERAL_POOL);
+            assertNotNull(generalPool);
+
+            // Wait for the cluster to run out of memory
+            while (generalPool.getAssignedQueries() != 1 || reservedPool.getAssignedQueries() != 1 || generalPool.getBlockedNodes() == 0 || reservedPool.getBlockedNodes() == 0) {
+                MILLISECONDS.sleep(10);
+            }
+
+            // Both queries should still be running
+            for (QueryInfo info : queryRunner.getCoordinator().getQueryManager().getAllQueryInfo()) {
+                assertFalse(info.getState().isDone());
+            }
+
+            // Wait for one of the queries to die
+            boolean queryDone = false;
+            while (!queryDone) {
+                for (QueryInfo info : queryRunner.getCoordinator().getQueryManager().getAllQueryInfo()) {
+                    if (info.getState().isDone()) {
+                        assertEquals(info.getErrorCode().getCode(), CLUSTER_OUT_OF_MEMORY.toErrorCode().getCode());
+                        queryDone = true;
+                        break;
+                    }
+                }
+                MILLISECONDS.sleep(10);
+            }
+
+            // Release the memory in the reserved pool
+            for (TestingPrestoServer server : queryRunner.getServers()) {
+                MemoryPool reserved = server.getLocalMemoryManager().getPool(RESERVED_POOL);
+                // Free up the entire pool
+                reserved.free(fakeQueryId, reserved.getMaxBytes());
+                assertTrue(reserved.getFreeBytes() > 0);
+            }
+
+            for (Future<?> query : queryFutures) {
+                query.get();
```

Comment:
Yep, it's running on tpch tiny, so it should finish quite quickly once the memory is free